0013-RS-pubsub-subscription-streaming.md

```diff
@@ -0,0 +1,195 @@
+  // can be processed by the subscriber at any given time.
+  // Default is no limit.
+  optional max_in_flight_messages = 5 [json_name = "maxInFlightMessages"];
+}
+
+// SubscribeTopicEventsResponse is a message containing the result of a
+// subscription to a topic.
+message SubscribeTopicEventsResponseAlpha1 {
+  // id is the unique identifier for the subscription request.
+  string id = 1 [json_name = "id"];
+
+  // status is the result of the subscription request.
+  TopicEventResponseAlpha1 status = 2 [json_name = "status"];
+}
+```
+
+When an application wishes to subscribe to a topic it will initiate a stream with `SubscribeTopicEventsRequest`, and `Send` the initial request `SubscribeTopicEventsSubscribeRequest` containing the options for the subscription.
+Daprd will then setup the machinery to add this gRPC RPC stream to the set of subscribers.
+The request contains no route or path matching configuration as all events will be sent on this stream.
+Subscription gRPC streams are the highest priority when Daprd determines which publisher a message should be sent to.
+Only a single PubSub Topic pair may be subscribed at a single time with this API.
+If the first message sent to the server is not the initial request, the RPC will return an error.
+If any subsequent messages are not `SubscribeTopicEventsResponse` messages, the RPC will return an error.
+
+When a message is published to the topic, Daprd will send a `TopicEventRequest` message on the stream containing the message payload and metadata.
+After the application has processed the message, it will send to the server a `SubscribeTopicEventsResponse` containing the `id` of the message and the `status` of the message processing.
+Since multiple messages can be sent and processed in the application at the same time, the event `id` is used by the server to track the status of each individual event.
+An event topic response will follow the timeout resiliency as currently exist for subscriptions.
+
+Client code:
+
+```go
+	stream, _ := client.SubscribeTopicEventsAlpha1(ctx)
+	stream.Send(&rtv1.SubscribeTopicEventsRequestAlpha1{
+		SubscribeTopicEventsRequestTypeAlpha1: &rtv1.SubscribeTopicEventsRequest_RequestAlpha1{
+			Request: &rtv1.SubscribeTopicEventsSubscribeRequestAlpha1{
+				PubsubName: "mypub", Topic: "a",
+			},
+		},
+	})
+
+	client.PublishEvent(ctx, &rtv1.PublishEventRequest{
+		PubsubName: "mypub", Topic: "a",
+		Data:            []byte(`{"status": "completed"}`),
+		DataContentType: "application/json",
+	})
+
+	event, _ := stream.Recv()
+	stream.Send(&rtv1.SubscribeTopicEventsRequestAlpha1{
+		SubscribeTopicEventsRequestType: &rtv1.SubscribeTopicEventsRequest_EventResponseAlpha1{
+			EventResponse: &rtv1.SubscribeTopicEventsResponseAlpha1{
+				Id:     event.Id,
+				Status: &rtv1.TopicEventResponse{Status: rtv1.TopicEventResponse_SUCCESS},
+			},
+		},
+	})
+
+	stream.CloseSend()
+```
+
+### HTTP (WebSockets)
```

Comment:
WebSockets make things a lot more complicated, including things like authz. We support HTTP/2 now so we should just use server-sent events (since we don't need bi-di streaming). It's much simpler to implement and use, and doesn't introduce a lot more dependencies on the server.

Reply:
We can't use unidirectional streaming because the client needs to send back information to the server (Dapr), which is the status for each message received. From the auth perspective this wouldn't be different from how an app authenticates to Dapr today (via an access token, and unauthenticated by default) and even in browser to server scenarios a client access token is used.

Reply:
The send-back can be done with regular HTTP requests. When using HTTP/2, they use the same TCP socket, so this wouldn't be an issue.

I would recommend avoiding adding another protocol as it does make a lot of things more complicated.

1. Go doesn't support WebSockets over HTTP/2, but only HTTP/1, so the app requires a separate socket: https://github.com/golang/go/issues/49918
2. When connecting to a WS using a web browser, you can't customize the headers, so Dapr authentication wouldn't be usable (https://devcenter.heroku.com/articles/websocket-security#authentication-authorization). I understand that connecting directly to Dapr through a web browser is not officially supported, but some customers do that. Hence why I mentioned auth is more challenging.
3. WebSockets also adds complexities when dealing with proxies (this may be a problem for some Diagrid services too presumably :) ), and local development requires tooling that is WebSockets-aware.

Reply:
The auth remains token based auth, but it doesn't have to pass through a header in this case. The initial handshake contains the token as described in the link you posted. This seems like a standard way for client side auth and is also used by the Azure Websockets (PubSub) service, as one example.

Re: web browser, it'd actually be great to make this HTTP path usable from web browsers, where http/2 doesn't have the same support as websockets. This will provide a clearer usage pattern where backend services can connect over gRPC while web services and/or any client that doesn't have gRPC or http/2 support can use websockets.

The only issue we face with websockets is the separate socket issue. If the same port cannot be shared with the existing HTTP server and the Websockets server than this is likely a non-starter.

Reply:
> The auth remains token based auth, but it doesn't have to pass through a header in this case. The initial handshake contains the token as described in the link you posted.

Correct, but that's part of the application protocol we need to define, and it does add complexity.

> it'd actually be great to make this HTTP path usable from web browsers, where http/2 doesn't have the same support as websockets

[Support for HTTP/2](https://caniuse.com/?search=http%2F2) in browser is just as much as [WebSockets](https://caniuse.com/?search=websockets). In any case, HTTP/2 is a nice-to-have for long-lived connections but not a requirement (using HTTP/1 would be fine, but would require one separate TCP socket).

> The only issue we face with websockets is the separate socket issue. If the same port cannot be shared with the existing HTTP server and the Websockets server than this is likely a non-starter.

Separate **socket**, not port. So the long-lived connection requires its own TCP socket between the app and Dapr (1 TCP socket for the WebSockets long-lived stream, and then each request from the app will have its own). While in the case of HTTP/2 there's multiplexing being used, so there is at most 1 TCP socket being used (the long-lived stream can share the same socket as every request going to Dapr). This does improve efficiency. But it doesn't require a separate port - the same Go server can serve HTTP/1, HTTP/2, and WebSockets (but this only over HTTP/1 and not HTTP/2, per issue above)

Reply:
Ok, so if we don't need a separate port and there's a path forward for HTTP/2 support for WebSockets then I'm comfortable using WebSockets HTTP 1.1 to start with

Reply:
The [issue for adding WebSockets over HTTP/2](https://github.com/golang/go/issues/49918) was opened in 2021 and looks like it's not being worked on. It seems that it is not a priority (understandably, as WebSockets is not as popular as it once was).

I think enabling a new communication protocol should be something not done lightly, as we'll be on the hook to support that for a very long time. There's nothing inherently wrong with HTTP/2, which is already supported and we have all tooling and test in place.

Reply:
The user experience that comes out of having a full-duplex bidi stream with the receive message/reply to server commands encapsulated within WebSockets is preferable to me than what the experience would be like with SSE, moreover the latter being limited to a text format. We can always change to HTTP2/SSE if we're seeing a reason to do so before moving the API to stable

Reply:
Although WebSocket is not that popular, it is still very useful in full-duplex bidi stream case. I think it is a good start and we can Iterate it based on users' response and so on.

Reply:
I am open to looking at the maintainability aspect of Websockets and we could also potentially ping for the interest in the duplex stream from the community. 
But this proposal looks good for gRPC as a starter.